pkg/resources/gce/gce.go

```diff
@@ -796,25 +799,80 @@ func (d *clusterDiscoveryGCE) matchesClusterNameMultipart(name string, maxParts
 	return false
 }
 
+func (d *clusterDiscoveryGCE) clusterDNSName() string {
+	return d.clusterName + "."
+}
+
+func (d *clusterDiscoveryGCE) isKopsManagedDNSName(name string) bool {
+	prefix := []string{`api`, `api.internal`, `bastion`}
+	for _, p := range prefix {
+		if name == p+"."+d.clusterDNSName() {
+			return true
+		}
+	}
+	return false
+}
+
 func (d *clusterDiscoveryGCE) listGCEDNSZone() ([]*resources.Resource, error) {
-	// We never delete the hosted zone, because it is usually shared and we don't create it
-	return nil, nil
-	// TODO: When shared resource PR lands, reintroduce
-	//if dns.IsGossipHostname(d.clusterName) {
-	//	return nil, nil
-	//}
-	//zone, err := d.findDNSZone()
-	//if err != nil {
-	//	return nil, err
-	//}
-	//
-	//return []*resources.Resource{
-	//	{
-	//		Name:    zone.Name(),
-	//		ID:      zone.Name(),
-	//		Type:    "DNS Zone",
-	//		Deleter: d.deleteDNSZone,
-	//		Obj:     zone,
-	//	},
-	//}, nil
+
+	if dns.IsGossipHostname(d.clusterName) {
+		return nil, nil
+	}
+
+	var resourceTrackers []*resources.Resource
+
+	zoneResponse, err := d.gceCloud.CloudDNS().ManagedZones.List(d.gceCloud.Project()).Do()
+	if err != nil {
+		return nil, fmt.Errorf("error getting GCE DNS zones %v", err)
+	}
+
+	for _, zone := range zoneResponse.ManagedZones {
+		if !strings.HasSuffix(d.clusterDNSName(), zone.DnsName) {
+			continue
+		}
+		response, err := d.gceCloud.CloudDNS().ResourceRecordSets.List(d.gceCloud.Project(), zone.Name).Do()
+		if err != nil {
+			return nil, fmt.Errorf("error getting GCE DNS zone data %v", err)
+		}
+
+		for _, record := range response.Rrsets {
+			// adapted from AWS implementation
+			if record.Type != "A" {
+				continue
+			}
+
+			if d.isKopsManagedDNSName(record.Name) {
+				resource := resources.Resource{
+					Name:         zone.Name,
```

Comment:
I think this is shown to the user, so it might be nice to keep this as record.Name